cookbook/security/form_login_setup.rst

```diff
@@ -165,39 +165,35 @@ form::
 
     // src/AppBundle/Controller/SecurityController.php
     // ...
-    
+
     // ADD THIS use STATEMENT above your class
     use Symfony\Component\Security\Core\Security;
 
     public function loginAction(Request $request)
     {
-        $session = $request->getSession();
+        $authenticationUtils = $this->get('security.authentication_utils');
 
         // get the login error if there is one
-        if ($request->attributes->has(Security::AUTHENTICATION_ERROR)) {
-            $error = $request->attributes->get(
-                Security::AUTHENTICATION_ERROR
-            );
-        } elseif (null !== $session && $session->has(Security::AUTHENTICATION_ERROR)) {
-            $error = $session->get(Security::AUTHENTICATION_ERROR);
-            $session->remove(Security::AUTHENTICATION_ERROR);
-        } else {
-            $error = '';
-        }
+        $error = $authenticationUtils->getLastAuthenticationError();
 
         // last username entered by the user
-        $lastUsername = (null === $session) ? '' : $session->get(Security::LAST_USERNAME);
+        $lastUsername = $authenticationUtils->getLastUsername();
 
         return $this->render(
             'security/login.html.twig',
             array(
                 // last username entered by the user
                 'last_username' => $lastUsername,
-                'error'         => $error,
+                'error'         => $error ?: '',
```

Comment:
no, don't call ``getMessage``. This is a security issue (it can leak your database structure if a PDO exception happens). You should use ``getMessageKey`` and ``getMessageParameters`` to render the AuthenticationException to the user (translating the message key in the ``security`` domain to get in the right locale).
But the template rendering the error should be fixed in the 2.3 branch, not here, because it applies to 2.3+

Btw, currently, the code works, because it calls getMessage in the template.